[aarch64][win] Add support for import call optimization (equivalent to MSVC /d2ImportCallOptimization)

dpaoliello opened the pull request:
This change implements import call optimization for AArch64 Windows (equivalent to the undocumented MSVC `/d2ImportCallOptimization` flag).

Import call optimization adds additional data to the binary which can be used by the Windows kernel loader to rewrite indirect calls to imported functions as direct calls. It uses the same [Dynamic Value Relocation Table mechanism that was leveraged on x64 to implement `/d2GuardRetpoline`](https://techcommunity.microsoft.com/blog/windowsosplatform/mitigating-spectre-variant-2-with-retpoline-on-windows/295618).

The change to the obj file is to add a new `.impcall` section with the following layout:
```cpp
  // Per section that contains calls to imported functions:
  //  uint32_t SectionSize: Size in bytes for information in this section.
  //  uint32_t Section Number
  //  Per call to imported function in section:
  //    uint32_t Kind: the kind of imported function.
  //    uint32_t BranchOffset: the offset of the branch instruction in its
  //                            parent section.
  //    uint32_t TargetSymbolId: the symbol id of the called function.
```

NOTE: If the import call optimization feature is enabled, then the `.impcall` section must be emitted, even if there are no calls to imported functions.

The implementation is split across a few parts of LLVM:
* During AArch64 instruction selection, the `GlobalValue` for each call to a global is recorded into the Extra Information for that node.
* During lowering to machine instructions, the called global value for each call is noted in its containing `MachineFunction`.
* During AArch64 asm printing, if the import call optimization feature is enabled:
  - A (new) `.impcall` directive is emitted for each call to an imported function.
  - The `.impcall` section is emitted with its magic header (but is not filled in).
* During COFF object writing, the `.impcall` section is filled in based on each `.impcall` directive that were encountered.

The `.impcall` section can only be filled in when we are writing the COFF object as it requires the actual section numbers, which are only assigned at that point (i.e., they don't exist during asm printing).

I had tried to avoid using the Extra Information during instruction selection and instead implement this either purely during asm printing or in a `MachineFunctionPass` (as suggested in [on the forums](https://discourse.llvm.org/t/design-gathering-locations-of-instructions-to-emit-into-a-section/83729/3)) but this was not possible due to how loading and calling an imported function works on AArch64. Specifically, they are emitted as `ADRP` + `LDR` (to load the symbol) then a `BR` (to do the call), so at the point when we have machine instructions, we would have to work backwards through the instructions to discover what is being called. An initial prototype did work by inspecting instructions; however, it didn't correctly handle the case where the same function was called twice in a row, which caused LLVM to elide the `ADRP` + `LDR` and reuse the previously loaded address. Worse than that, sometimes for the double-call case LLVM decided to spill the loaded address to the stack and then reload it before making the second call. So, instead of trying to implement logic to discover where the value in a register came from, I instead recorded the symbol being called at the last place where it was easy to do: instruction selection.

dpaoliello committed:
[aarch64][win] Add support for import call optimization (equivalent to MSVC /d2ImportCallOptimization)

arsenm commented on llvm/lib/MC/MCParser/COFFAsmParser.cpp:
```diff
@@ -577,6 +580,24 @@ bool COFFAsmParser::parseDirectiveSymIdx(StringRef, SMLoc) {
 
+bool COFFAsmParser::parseDirectiveImpCall(StringRef, SMLoc) {
+  if (!getContext().getObjectFileInfo()->getImportCallSection())
+    return TokError("target doesn't have an import call section");
```
Avoid ' in error messages

arsenm commented on llvm/test/MC/AArch64/win-import-call-optimization.s:
```diff
@@ -0,0 +1,62 @@
+// CHECK-NEXT:       0x[[#%x,TCOFFSET - 8]] IMAGE_REL_ARM64_PAGEBASE_REL21 __imp_b ([[#%u,TCSYM]])
+// CHECK-NEXT:       0x[[#%x,TCOFFSET - 4]] IMAGE_REL_ARM64_PAGEOFFSET_12L __imp_b ([[#%u,TCSYM]])
+// CHECK-NEXT:     }
+// CHECK-NEXT:   ]
```
End of file whitespace error

arsenm commented on llvm/lib/MC/MCParser/COFFAsmParser.cpp:
```diff
@@ -577,6 +580,24 @@ bool COFFAsmParser::parseDirectiveSymIdx(StringRef, SMLoc) {
+    return TokError("target doesn't have an import call section");
+
+  StringRef SymbolID;
+  if (getParser().parseIdentifier(SymbolID))
+    return TokError("expected identifier in directive");
+
+  if (getLexer().isNot(AsmToken::EndOfStatement))
+    return TokError("unexpected token in directive");
```
Don't think these error cases are tested

arsenm commented on llvm/test/MC/AArch64/win-import-call-optimization-no-section.s:
```diff
@@ -0,0 +1,9 @@
+  .impcall        __imp_b
+  br     x8
+
+// CHECK: error: .impcall directives were used, but no existing .impcall section exists
```
Is it possible to include a source location

dpaoliello replied:
Unfortunately, no: the `.impcall` section may appear after the `.impcall` directives, so I'd need to keep track of the locations of all (or at least one) observed `.impcall` directives, which is not easy to do with the way asm parsing vs object writing is setup.

mstorsjo commented:
A couple questions on the mechanism:
- As the flag `/d2ImportCallOptimization` is undocumented, I presume that this format is undocumnted too. This covers two separate formats as far as I can see - the `.impcall` section contents, which this PR generates, and which the linker consumes. This format is mostly a convention - an agreement between compiler and linker; this currently uses the `Imp_Call_V1` identifier. And then secondly, the dynamic relocations that the linker generates, which ends up in the final binary, which is consumed by the windows loader. This format isn't touched upon here (as this only covers code generation, not linking). The format of those dynamic relocations is much more fixed, as it's handled by the OS. Although this is only an optimization, so running on older Windows versions which doesn't recognize it, should be fine too?
- When the linker uses these dynamic relocations, it changes a `br` or `blr` indirect branch instruction into a direct `b` or `bl`, if the target address is close enough within the address space - right? (The other instructions for loading the address are left untouched, as those instructions can be anywhere detached from the branch, and we can't speculate on whether the register contents is needed elsewhere.)
- As the 64 bit address space is kinda large, and the `b` and `bl` instructions only have a +/- 128 MB range, I guess that this optimization simply can't be applied, if the target is too far away? Intuitively, it feels like this wouldn't end up effective all that often? Then again, I guess most DLLs are loaded close to each other in the virtual memory layout, and the base EXE, with dynamic base enabled (as always on aarch64) also would end up somewhere close.
- Is the only thing we gain here, the performance benefits of a direct branch, which is easier for the branch predictor, compared to an indirect branch? While that obviously is better, this feels like a whole lot of extra work and structures, for something that feels like relatively small gains. Is there something else to be gained in relation to this as well that I'm missing (and/or that isn't mentioned yet)? Or is the gains from better branch prediction much bigger than what I'm thinking here?
- For cases when dll imported functions are called without being marked as dllimport in headers, we jump via a thunk (from the import library, and/or linker generated). Can the same thing be applied to them? Does that require including similar `.impcall` sections in the import libraries for the thunks? (In the case of lld, it actually doesn't use the import library contents here but just synthesize it on its own - there we could do the same right away without needing to update import libraries with this metadata.)

dpaoliello commented:
> As the flag /d2ImportCallOptimization is undocumented, I presume that this format is undocumnted too. This covers two separate formats as far as I can see - the .impcall section contents, which this PR generates, and which the linker consumes. This format is mostly a convention - an agreement between compiler and linker; this currently uses the Imp_Call_V1 identifier. And then secondly, the dynamic relocations that the linker generates, which ends up in the final binary, which is consumed by the windows loader. This format isn't touched upon here (as this only covers code generation, not linking). The format of those dynamic relocations is much more fixed, as it's handled by the OS. Although this is only an optimization, so running on older Windows versions which doesn't recognize it, should be fine too?

Correct on all accounts:
* This PR only adds compiler support, with no linker support.
* The format of this section is only used by the linker and is not present in the final binary.
* The linker uses this to build the "Dynamic Value Relocation Table", and [older versions of Windows will ignore entries in that table it does not understand](https://techcommunity.microsoft.com/blog/windowsosplatform/mitigating-spectre-variant-2-with-retpoline-on-windows/295618#:~:text=Importantly%2C%20the%20kernel%20will%20skip%20over%20any%20DVRT%20entries%20it%20does%20not%20recognize).

dpaoliello commented:
> When the linker uses these dynamic relocations, it changes a br or blr indirect branch instruction into a direct b or bl, if the target address is close enough within the address space - right? (The other instructions for loading the address are left untouched, as those instructions can be anywhere detached from the branch, and we can't speculate on whether the register contents is needed elsewhere.)
> As the 64 bit address space is kinda large, and the b and bl instructions only have a +/- 128 MB range, I guess that this optimization simply can't be applied, if the target is too far away? Intuitively, it feels like this wouldn't end up effective all that often? Then again, I guess most DLLs are loaded close to each other in the virtual memory layout, and the base EXE, with dynamic base enabled (as always on aarch64) also would end up somewhere close.

The linker doesn't do the rewriting, the kernel-mode loader does, but yes it will rewrite the indirect branches to direct branches. Since it is the loader doing this work, it knows how far away the target address is and, therefore, if the instruction can be rewritten or not. Since this is an optimization, the loader can choose not to rewrite the instruction without affecting correctness.

dpaoliello commented:
> Is the only thing we gain here, the performance benefits of a direct branch, which is easier for the branch predictor, compared to an indirect branch? While that obviously is better, this feels like a whole lot of extra work and structures, for something that feels like relatively small gains. Is there something else to be gained in relation to this as well that I'm missing (and/or that isn't mentioned yet)? Or is the gains from better branch prediction much bigger than what I'm thinking here?

I don't have the exact numbers, but we did see significant performance improvement within the Windows kernel for some components (which were especially chatty with other .sys files) when MSVC added this optimization. I'll see if I can get one of my colleagues to provide more details.

dpaoliello commented:
> For cases when dll imported functions are called without being marked as dllimport in headers, we jump via a thunk (from the import library, and/or linker generated). Can the same thing be applied to them? Does that require including similar .impcall sections in the import libraries for the thunks? (In the case of lld, it actually doesn't use the import library contents here but just synthesize it on its own - there we could do the same right away without needing to update import libraries with this metadata.)

I'm not 100% sure how MSVC handles this - there are no changes to import libraries (i.e., there's no `.impcall` section), so I'm guessing it would be done within the linker.

If lld supports generating the Dynamic Value Relocation Table, there's no reason it can't add additional entries in there for other indirect branches to imported functions that it knows about.

dpaoliello commented:
> > Is the only thing we gain here, the performance benefits of a direct branch, which is easier for the branch predictor, compared to an indirect branch? While that obviously is better, this feels like a whole lot of extra work and structures, for something that feels like relatively small gains. Is there something else to be gained in relation to this as well that I'm missing (and/or that isn't mentioned yet)? Or is the gains from better branch prediction much bigger than what I'm thinking here?
> 
> I don't have the exact numbers, but we did see significant performance improvement within the Windows kernel for some components (which were especially chatty with other .sys files) when MSVC added this optimization. I'll see if I can get one of my colleagues to provide more details.

I have some numbers: when applied to the Arm64 Windows Kernel, we saw a 3% throughput improvement on DiskSpd benchmarks for both SnapDragon 8cx Gen2 and Gen3 devices.

mstorsjo commented:
> I have some numbers: when applied to the Arm64 Windows Kernel, we saw a 3% throughput improvement on DiskSpd benchmarks for both SnapDragon 8cx Gen2 and Gen3 devices.

Interesting - that's a massive number a change like this. I'm not too familiar with the structure of the kernel though, I presume that there would need to be a huge number of cross-DLL function calls within the kernel for this change to make any measurable difference.

efriedma-quic reviewed (commented):
The mechanism for tracking the call from ISel through the assembler seems generally fine.  Alternatively, you could add an operand to call instructions, but I don't see any reason to prefer that.

efriedma-quic commented:
> The .impcall section can only be filled in when we are writing the COFF object as it requires the actual section numbers, which are only assigned at that point (i.e., they don't exist during asm printing).

You could possibly add a directive specifically to emit section numbers... that would allow you to move more of the logic into the AsmPrinter.  We already have `.secidx`; that's not quite what you want here, but maybe to give you some idea what it would look like.

Not sure if that's worth doing.

dpaoliello commented:
> > The .impcall section can only be filled in when we are writing the COFF object as it requires the actual section numbers, which are only assigned at that point (i.e., they don't exist during asm printing).
> 
> You could possibly add a directive specifically to emit section numbers... that would allow you to move more of the logic into the AsmPrinter. We already have `.secidx`; that's not quite what you want here, but maybe to give you some idea what it would look like.

I considered this for an earlier design, but it ended up being very messy: the section numbers are assigned after the sections have been streamed, so if there is a general purpose "insert section number here" mechanism, it needs to be able to go back and rewrite already-streamed sections. The current design avoids that complexity since we only use the section numbers in the `.impcall` sections which we pre-allocate but then fill in afterwards.

dpaoliello commented on llvm/lib/MC/WinCOFFObjectWriter.cpp:
```diff
@@ -1143,6 +1167,51 @@ uint64_t WinCOFFWriter::writeObject(MCAssembler &Asm) {
+void WinCOFFWriter::recordImportCall(const MCDataFragment &FB,
+                                     const MCSymbol *Symbol) {
+  auto &SectionData = SectionToImportCallsMap[FB.getParent()];
+  SectionData.push_back(ImportCall{unsigned(FB.getContents().size()), Symbol});
```
Good call - I switched this to using symbols and `Asm.getSymbolOffset`

efriedma-quic commented:
> it needs to be able to go back and rewrite already-streamed sections

Yes, a relocation (at the assembler level, not in the resulting object).  We have infrastructure to do this if you wanted to.

dpaoliello commented:
> Yes, a relocation (at the assembler level, not in the resulting object). We have infrastructure to do this if you wanted to.

Sure, can you point me to the code?

I'd be happy to get rid of the `.impcall` directive and do most of the work in the asm printer, as it'll make the x86 implementation easier.

efriedma-quic commented:
Internally it should be an MCFixup.  I guess you'd add a new fixup kind.

Maybe follow the code for `.secidx` to get some idea what that would look like. It's not quite the same, because you need to resolve it instead of emitting a relocation, but before that point it's similar.

dpaoliello commented:
Ok, switched this over to using section number and offset assembly directives, implemented using new `MCTargetExpr` derived types (which avoided the need to introduce any new fixup kinds and localizes the changes to just the COFF streamer and object writer).

This also meant that writing the `.impcall` section is now completely contained in the asm printer, avoiding the need for `.impcall` directives (and the weird failure cases associated with them) and will make the x86 implementation much easier.

dpaoliello commented on llvm/include/llvm/CodeGen/MachineFunction.h:
```diff
@@ -354,6 +354,11 @@ class LLVM_ABI MachineFunction {
+  /// Mapping of call instruction to the global value and target flags that it
+  /// calls, if applicable.
+  DenseMap<const MachineInstr *, std::pair<const GlobalValue *, unsigned>>
+      CalledGlobalsMap;
```
Added. Are there explicit tests for this?
(I ran into some crashes while testing that I fixed, so I know that round-tripping is working...)

arsenm replied:
test/CodeGen/MIR is for print/parse tests

dpaoliello replied:
Excellent, thanks! I've added a test.

efriedma-quic reviewed (approved):
LGTM

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -77,6 +78,11 @@ static cl::opt<PtrauthCheckMode> PtrauthAuthChecks(
     cl::init(Default));
 
+static cl::opt<bool> EnableImportCallOptimization(
+    "aarch64-win-import-call-optimization", cl::Hidden,
```
Oops, one thing I forgot to ask about.  How do we enable this?  An LLVM flag is fine for now, but we don't like to use them as the primary mechanism for enabling code generation options.  It doesn't interact well with LTO, or programmatically enabling the option.

We usually use function attributes for options which make sense to enable on a per-function basis.

You can address this in a followup.

dpaoliello replied:
I can do a follow-up PR to switch this to a function attribute (again, it'll make x64 easier since I'll probably have to disable some optimizations to get it to work).
My only question with that is: is there a way to mark all functions with an attribute without having to add an option to the frontend? Or am I going to have to flow this through to Clang?

efriedma-quic replied:
Yes, you'll have to make the clang driver pass this down as a clang codegen option (clang/include/clang/Basic/CodeGenOptions.def), then make clang codegen add the attribute.  This sort of thing is pretty common.

fhahn commented:
It looks like the change introduced a use-after-poison https://lab.llvm.org/buildbot/#/builders/169/builds/7294/steps/10/logs/stdio

```
FAIL: LLVM :: CodeGen/AArch64/machine-outliner-throw.ll (1 of 88242)
******************** TEST 'LLVM :: CodeGen/AArch64/machine-outliner-throw.ll' FAILED ********************
Exit Code: 2
Command Output (stderr):
--
RUN: at line 1: /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -verify-machineinstrs -enable-machine-outliner -mtriple=aarch64 < /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll | /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll
+ /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll
+ /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -verify-machineinstrs -enable-machine-outliner -mtriple=aarch64
RUN: at line 2: /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -verify-machineinstrs -enable-machine-outliner -mtriple=aarch64 -stop-after=machine-outliner < /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll | /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll -check-prefix=TARGET_FEATURES
+ /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/FileCheck /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/test/CodeGen/AArch64/machine-outliner-throw.ll -check-prefix=TARGET_FEATURES
+ /home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc -verify-machineinstrs -enable-machine-outliner -mtriple=aarch64 -stop-after=machine-outliner
=================================================================
==1112042==ERROR: AddressSanitizer: use-after-poison on address 0x52100002dc30 at pc 0x5979d93c2be3 bp 0x7ffd78825d90 sp 0x7ffd78825d88
READ of size 8 at 0x52100002dc30 thread T0
    #0 0x5979d93c2be2 in getParent /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/CodeGen/MachineInstr.h:347:55
    #1 0x5979d93c2be2 in llvm::MIRPrinter::convertCalledGlobals(llvm::yaml::MachineFunction&, llvm::MachineFunction const&, llvm::MachineModuleSlotTracker&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MIRPrinter.cpp:609:25
    #2 0x5979d93b74ab in llvm::MIRPrinter::print(llvm::MachineFunction const&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MIRPrinter.cpp:275:3
    #3 0x5979d93cd392 in llvm::printMIR(llvm::raw_ostream&, llvm::MachineModuleInfo const&, llvm::MachineFunction const&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MIRPrinter.cpp:1071:11
    #4 0x5979d9417ede in (anonymous namespace)::MIRPrintingPass::runOnMachineFunction(llvm::MachineFunction&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MIRPrintingPass.cpp:65:5
    #5 0x5979d8fdef61 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:13
    #6 0x5979d9e9e183 in llvm::FPPassManager::runOnFunction(llvm::Function&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:27
    #7 0x5979d9eb523e in llvm::FPPassManager::runOnModule(llvm::Module&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:16
    #8 0x5979d9e9fd7a in runOnModule /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:27
    #9 0x5979d9e9fd7a in llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:44
    #10 0x5979d408e2dc in compileModule(char**, llvm::LLVMContext&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/tools/llc/llc.cpp:751:8
    #11 0x5979d408882f in main /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/tools/llc/llc.cpp:411:22
    #12 0x7cb6a4a2a3b7  (/lib/x86_64-linux-gnu/libc.so.6+0x2a3b7) (BuildId: 5f3f024b472f38389da3a2f567b3d0eaa8835ca2)
    #13 0x7cb6a4a2a47a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a47a) (BuildId: 5f3f024b472f38389da3a2f567b3d0eaa8835ca2)
    #14 0x5979d3f9c0a4 in _start (/home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc+0xb4bf0a4)
0x52100002dc30 is located 2864 bytes inside of 4096-byte region [0x52100002d100,0x52100002e100)
allocated by thread T0 here:
    #0 0x5979d407bd82 in operator new(unsigned long, std::align_val_t) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/compiler-rt/lib/asan/asan_new_delete.cpp:98:3
    #1 0x5979db8b4e3d in llvm::allocate_buffer(unsigned long, unsigned long) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/Support/MemAlloc.cpp:16:10
    #2 0x5979d4516e59 in Allocate /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/Support/AllocatorBase.h:92:12
    #3 0x5979d4516e59 in StartNewSlab /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/Support/Allocator.h:346:42
    #4 0x5979d4516e59 in llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096ul, 4096ul, 128ul>::AllocateSlow(unsigned long, unsigned long, llvm::Align) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/Support/Allocator.h:202:5
    #5 0x5979d8ff9ed8 in Allocate /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/Support/Allocator.h:216:12
    #6 0x5979d8ff9ed8 in allocate<llvm::BumpPtrAllocatorImpl<llvm::MallocAllocator, 4096UL, 4096UL, 128UL> > /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/Support/ArrayRecycler.h:130:38
    #7 0x5979d8ff9ed8 in allocateOperandArray /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/CodeGen/MachineFunction.h:1117:28
    #8 0x5979d8ff9ed8 in llvm::MachineInstr::MachineInstr(llvm::MachineFunction&, llvm::MCInstrDesc const&, llvm::DebugLoc, bool) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MachineInstr.cpp:108:19
    #9 0x5979d8fb74cc in llvm::MachineFunction::CreateMachineInstr(llvm::MCInstrDesc const&, llvm::DebugLoc, bool) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MachineFunction.cpp:433:7
    #10 0x5979d45d7a8c in llvm::BuildMI(llvm::MachineFunction&, llvm::MIMetadata const&, llvm::MCInstrDesc const&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/CodeGen/MachineInstrBuilder.h:375:37
    #11 0x5979db2175cd in llvm::InstrEmitter::EmitMachineNode(llvm::SDNode*, bool, bool, llvm::SmallDenseMap<llvm::SDValue, llvm::Register, 16u, llvm::DenseMapInfo<llvm::SDValue, void>, llvm::detail::DenseMapPair<llvm::SDValue, llvm::Register>>&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/InstrEmitter.cpp:1060:29
    #12 0x5979db25acab in EmitNode /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/InstrEmitter.h:145:7
    #13 0x5979db25acab in llvm::ScheduleDAGSDNodes::EmitSchedule(llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>&)::$_0::operator()(llvm::SDNode*, bool, bool, llvm::SmallDenseMap<llvm::SDValue, llvm::Register, 16u, llvm::DenseMapInfo<llvm::SDValue, void>, llvm::detail::DenseMapPair<llvm::SDValue, llvm::Register>>&) const /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/ScheduleDAGSDNodes.cpp:874:13
    #14 0x5979db2587c3 in llvm::ScheduleDAGSDNodes::EmitSchedule(llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/ScheduleDAGSDNodes.cpp:965:9
    #15 0x5979db494dbe in llvm::SelectionDAGISel::CodeGenAndEmitDAG() /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1151:42
    #16 0x5979db48d112 in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1898:7
    #17 0x5979db4848d1 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:615:3
    #18 0x5979db47dd38 in llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:375:20
    #19 0x5979d8fdef61 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:13
    #20 0x5979d9e9e183 in llvm::FPPassManager::runOnFunction(llvm::Function&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:27
    #21 0x5979d9eb523e in llvm::FPPassManager::runOnModule(llvm::Module&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:16
    #22 0x5979d9e9fd7a in runOnModule /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:27
    #23 0x5979d9e9fd7a in llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:44
    #24 0x5979d408e2dc in compileModule(char**, llvm::LLVMContext&) /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/tools/llc/llc.cpp:751:8
    #25 0x5979d408882f in main /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/tools/llc/llc.cpp:411:22
    #26 0x7cb6a4a2a3b7  (/lib/x86_64-linux-gnu/libc.so.6+0x2a3b7) (BuildId: 5f3f024b472f38389da3a2f567b3d0eaa8835ca2)
    #27 0x7cb6a4a2a47a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a47a) (BuildId: 5f3f024b472f38389da3a2f567b3d0eaa8835ca2)
    #28 0x5979d3f9c0a4 in _start (/home/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/llc+0xb4bf0a4)
SUMMARY: AddressSanitizer: use-after-poison /home/b/sanitizer-x86_64-linux-fast/build/llvm-project/llvm/include/llvm/CodeGen/MachineInstr.h:347:55 in getParent
Shadow bytes around the buggy address:
  0x52100002d980: 00 00 00 00 00 00 f7 f7 f7 f7 f7 f7 f7 f7 f7 f7
  0x52100002da00: f7 00 00 00 00 00 00 00 00 f7 f7 f7 f7 f7 f7 f7
  0x52100002da80: f7 f7 f7 f7 00 00 00 00 00 00 00 00 f7 f7 f7 f7
  0x52100002db00: f7 f7 f7 f7 f7 f7 f7 00 00 00 00 00 00 00 00 f7
  0x52100002db80: 00 00 00 00 00 00 00 00 00 f7 00 00 00 00 00 00
=>0x52100002dc00: 00 00 f7 f7 f7 f7[f7]f7 f7 f7 f7 f7 f7 00 00 00
  0x52100002dc80: 00 00 00 00 00 00 00 00 00 00 00 00 00 f7 00 00
  0x52100002dd00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x52100002dd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f7 00
  0x52100002de00: 00 00 00 00 00 00 00 00 f7 f7 f7 f7 f7 f7 f7 f7
  0x52100002de80: f7 f7 f7 00 00 00 00 00 00 00 00 f7 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1112042==ABORTING
```

Please take a look and revert the change if it takes longer to fix.

dpaoliello commented:
> It looks like the change introduced a use-after-poison

Investigating...

dpaoliello commented:
> > > It looks like the change introduced a use-after-poison
> > 
> > 
> > Investigating...
> 
> I found the root cause: even though I modeled "Called Globals" after "Call Site Info", I didn't all the move/copy/erase functions that keep the Calle Site Info in-sync with changes to the instructions. It should be an easy fix since I can rename and update the existing functions.

PR with fix: #122762

kstoimenov commented:
@dpaoliello please note this was reverted in 2f7ade4b5e399962e18f5f9a0ab0b7335deece51.
